pkg/syncer/cnsoperator/controller/cnsnodevmattachment/cnsnodevmattachment_controller.go

```diff
@@ -213,6 +213,40 @@ func (r *ReconcileCnsNodeVMAttachment) Reconcile(ctx context.Context,
 		// If the CnsNodeVMAttachment instance is already attached and
 		// not deleted by the user, remove the instance from the queue.
 		if instance.Status.Attached && instance.DeletionTimestamp == nil {
+			// This is an upgrade scenarion : In summary, we fetch the SV PVC and check if the
+			// CNS PVC protection finalizer exist. If the finalizer does not exist, it incurs that
+			// the attachment object was created with an older CSI. Hence we add the
+			// CNS PVC protection finalizer on the SV PVC in the current reconciliation loop.
+			pvc := &v1.PersistentVolumeClaim{}
+			err = r.client.Get(ctx, k8stypes.NamespacedName{Name: instance.Spec.VolumeName, Namespace: instance.Namespace}, pvc)
+			if err != nil {
+				msg := fmt.Sprintf("failed to get PVC with volumename: %q on namespace: %q. Err: %+v",
+					instance.Spec.VolumeName, instance.Namespace, err)
+				recordEvent(ctx, r, instance, v1.EventTypeWarning, msg)
+				return reconcile.Result{RequeueAfter: timeout}, csifault.CSIApiServerOperationFault, nil
+			}
+			cnsPvcFinalizerExists := false
+			// Check if cnsPvcFinalizerExists already exists.
+			for _, finalizer := range pvc.Finalizers {
+				if finalizer == cnsoperatortypes.CNSPvcFinalizer {
+					cnsPvcFinalizerExists = true
+					break
```

Comment:
nit: May be an info log here that finalizer already exists? Should be helpful while debugging.